### ISSUE TITLE
Unable to run parallel tests due to port conflict

### ISSUE THREAD

**Issue by emzyme20** [docker-archive-public/testcontainers.testcontainers-dotnet-legacy#54]
**Describe the bug**
When you construct the DatabaseContainerBuilder, you must specify exposed postgres port. Is there a way to use the .NET port of TestContainers against a postgres instance that configures it to use generic ports so that we can run unit tests in parallel and not have clashes with docker and postgres? 

I am continually seeing errors about port 5432 being in use.

There are some similar issues that were raised on the Java TestContainers project but your port doesn't seem to include the methods that allow you to get the connection string with the random port that was created.

This is the current setup code:

```
this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();
```

However, the specifying of the port breaks unit tests.

**To Reproduce**
1. Create more than one unit test 
2. Attempt to run them in parallel using something like NCrunch
3. See error 'Docker.DotNet.DockerApiException: Docker API responded with status code=InternalServerError, response={"message":"driver failed programming external connectivity on endpoint romantic_hofstadter (ad7dd3bb69f9011e18bd2ff3a9c10108de12e355d3a32570c14ea24edacc2580): Bind for 0.0.0.0:5432 failed: port is already allocated"}'

**Expected behavior**
The PostgreSqlContainer should be able to work with random ports being specified:

```
this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();
```

If you don't specify the port, the test times out. 

Links to similar issues on Java TestContainers page:

https://github.com/testcontainers/testcontainers-java/issues/482
https://github.com/testcontainers/testcontainers-java/issues/256

**Comment by HofmeisterAn** [comment]
Yes, e. g. `WithPortBindings((5432, 5550))` should work. This will expose the PostgreSQL port `5432` of the container and maps it to the host port `5550`. You can map multiple host ports to the same port of different containers.

**Comment by isen-ng** [comment]
Right now, it cannot automatically publish exposed ports to higher ports because auto publish to higher ports is not turned on.

What you can do is randomly assign an open port to each container with `WithPortBindings((5432, myRandomPort))` like @HofmeisterAn stated. I know that this method is not fool proof because the port may get used between checking if the port is free and actually using the port, but this is the easiest way out before the TestContainer-dotnet code can be fixed and it should work most of the time.

Here's the code to get a free port (extracted from https://stackoverflow.com/a/150974):

```
static int FreeTcpPort()
{
  TcpListener l = new TcpListener(IPAddress.Loopback, 0);
  l.Start();
  int port = ((IPEndPoint)l.LocalEndpoint).Port;
  l.Stop();
  return port;
}
```

**Comment by emzyme20** [comment]
Thanks, it now seems to be able to create the containers but is now timing out by not returning from the method that calls `Container.Start()`. The output from docker ps is:

![image](https://user-images.githubusercontent.com/3832400/59340679-5b214680-8cfe-11e9-9edf-29e6253822a8.png)

I've tried it with the helper method but obviously this is generating very high port numbers and you stated that it was not supported yet. I then modified it to generate a random number with a range and that is also timing out..

Can you tell me what range of port numbers it would support? 

```
            // var randomPort = FreeTcpPort();
            var randomPort = this.randomPortGenerator.Next(4000, 5000);
            this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                //.WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithPortBindings((5432, randomPort))
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();
```

The exception in the unit test is:

System.Exception: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout.
   at TestContainers.Core.Containers.PostgreSqlContainer.<WaitUntilContainerStarted>d__16.MoveNext()

**Comment by HofmeisterAn** [comment]
I'm a bit in hurry, and I just looked at it for a few second, but I guess you need to override the `ConnectionString` property and add your `randomPort` port to it. It looks like, `WaitUntilContainerStarted` does not consider custom ports so far.

https://github.com/testcontainers/testcontainers-dotnet/blob/5071119edcaf9a4c4633400991e5002885af65d8/src/TestContainers/Core/Containers/PostgreSqlContainer.cs#L25

**Comment by emzyme20** [comment]
Ok, should it be Port={randomPort} or should the port be appended to the host e.g. localhost:{randomPort}? or Should it have the custom port and 5432 defined somehow e.g. Port={randomPort}:5432?

This setup code still times out and I have tried various different ways to define the port
``` csharp
        public PostgreSqlFixture()
        {
            this.randomContainerPort = this.randomPortGenerator.Next(4000, 5000);
            this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithPortBindings((5432, this.randomContainerPort))
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();

            var connectionStringWithPort = this.ConnectionString;
            this.DocumentStore = global::Marten.DocumentStore.For(
                config =>
                {
                    config.Connection(connectionStringWithPort);
                });
        }

        public string ConnectionString => $"Host={this.Container.GetDockerHostIpAddress()};Port={this.randomContainerPort};Username={this.Container.UserName};pwd={this.Container.Password}";

```

**Comment by isen-ng** [comment]
As a start, you will need both `ExposedPorts` and `PortBindings`. This is just how docker works.

```
            this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithPortBindings((PostgreSqlContainer.POSTGRESQL_PORT, this.randomContainerPort))
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();
```

unfortunately this is incorrect:

```
public string ConnectionString => $"Host={this.Container.GetDockerHostIpAddress()};Port={this.randomContainerPort};Username={this.Container.UserName};pwd={this.Container.Password}";
```

It should be this because `this.Container.Password` will always return "Password123" no matter what password you set in the env variable. Same for the user if you are going to change the user.

```
string myPassword = "abcd1234";

this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithPortBindings((PostgreSqlContainer.POSTGRESQL_PORT, this.randomContainerPort))
                .WithEnv(("POSTGRES_PASSWORD", myPassword))
                .Build();

public string ConnectionString => $"Host={this.Container.GetDockerHostIpAddress()};Port={this.randomContainerPort};Username={this.Container.UserName};pwd={myPassword}";
```

Regarding the timeout policy, may I get the full stacktrace from that exception? or is this all?

**Comment by HofmeisterAn** [comment]
For PostgreSQL I use `$"Server={this.Hostname};Port={this.Port};Database={this.Database};User Id={this.Username};Password={this.Password};"`. Shouldn't make a big difference, I guess. [Here](https://www.connectionstrings.com/) you will find a collection about various connections strings.

I your case you have to use `randomContainerPort`, which should be better called `hostPort` or something like that.

@isen-ng FYI `.WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)` is not necessary. `WithPortBindings` will automatically expose the port via Docker API.

**Comment by emzyme20** [comment]
Thanks for the link to the connection strings, very useful.

The generated connection string in one of my tests generated the following:

```
Server=localhost;Port=4818;Database=TestDb;User Id=root;Password=Password123;
```

Using this setup, (N.B. I have moved the password and other variables to be defined by me and now call the extra methods on the `PostgreSqlContainer` to set them). I've tried it with `.WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)` included and also commented out (both fail)

``` csharp
        private readonly int hostPort;
        private const string UserName = "root";
        private const string DatabaseName = "TestDb";
        private readonly Random randomPortGenerator = new Random(DateTime.Now.Millisecond);
        private readonly string password;

        public PostgreSqlFixture()
        {
            this.password = "Password123";
            this.hostPort = this.randomPortGenerator.Next(4000, 5000);
            this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithPortBindings((5432, this.hostPort))
                .WithEnv(("POSTGRES_PASSWORD", this.password))
                .WithDatabaseName(DatabaseName)
                .WithUserName(UserName)
                .WithPassword(this.password)
                .Build();

            var connectionStringWithPort = this.ConnectionString;
            this.DocumentStore = global::Marten.DocumentStore.For(
                config =>
                {
                    config.Connection(connectionStringWithPort);
                });
        }

        public string ConnectionString =>
            $"Server={this.Container.GetDockerHostIpAddress()};Port={this.hostPort};Database={DatabaseName};User Id={UserName};Password={this.password};";
```

The full stack trace for the exception that I am getting is:

```
System.Exception : The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout.
---- Polly.Timeout.TimeoutRejectedException : The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout.
-------- System.Threading.Tasks.TaskCanceledException : A task was canceled.
   at TestContainers.Core.Containers.PostgreSqlContainer.<WaitUntilContainerStarted>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at TestContainers.Core.Containers.Container.<TryStart>d__39.MoveNext() in /home/appveyor/projects/testcontainers-dotnet-ak6r1/TestContainers/Core/Containers/Container.cs:line 58
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at TestContainers.Core.Containers.Container.<Start>d__38.MoveNext() in /home/appveyor/projects/testcontainers-dotnet-ak6r1/TestContainers/Core/Containers/Container.cs:line 33
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
----- Inner Stack Trace -----
   at Polly.Timeout.TimeoutEngine.<ImplementationAsync>d__1`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\TimeoutEngineAsync.cs:line 62
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Policy.<ExecuteAsyncInternal>d__180.MoveNext() in C:\projects\polly\src\Polly.Shared\Policy.Async.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Wrap.PolicyWrapEngine.<ImplementationAsync>d__9.MoveNext() in C:\projects\polly\src\Polly.Shared\Wrap\PolicyWrapEngineAsync.cs:line 101
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Policy.<ExecuteAsyncInternal>d__180.MoveNext() in C:\projects\polly\src\Polly.Shared\Policy.Async.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Policy.<ExecuteAndCaptureAsync>d__130.MoveNext() in C:\projects\polly\src\Polly.Shared\Policy.Async.ExecuteOverloads.cs:line 378
----- Inner Stack Trace -----
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Retry.RetryStateWaitAndRetryForever`1.<CanRetryAsync>d__8.MoveNext() in C:\projects\polly\src\Polly.Shared\Retry\RetryStateWaitAndRetryForeverAsync.cs:line 30
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Retry.RetryEngine.<ImplementationAsync>d__1`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Retry\RetryEngineAsync.cs:line 57
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Policy.<ExecuteAsyncInternal>d__180.MoveNext() in C:\projects\polly\src\Polly.Shared\Policy.Async.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Wrap.PolicyWrapEngine.<>c__DisplayClass9_0.<<ImplementationAsync>b__0>d.MoveNext() in C:\projects\polly\src\Polly.Shared\Wrap\PolicyWrapEngineAsync.cs:line 102
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Policy.<>c__DisplayClass213_1.<<TimeoutAsync>b__1>d.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\TimeoutSyntaxAsync.cs:line 256
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.TimeoutEngine.<ImplementationAsync>d__1`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\TimeoutEngineAsync.cs:line 34
```

**Comment by HofmeisterAn** [comment]
I doubt that another database name or user name expect the default values (`postgres`) are supported right now. Those values are set via environment variables like `POSTGRES_DB` and `POSTGRES_USER`, I cannot find any of them. You'll find further information [here](https://github.com/docker-library/docs/tree/master/postgres#environment-variables). @swissarmykirpan do you know more?

As soon as we finished the ci milestone, I can have a look at this. I think some functions are simply missing. In the meanwhile you can have a look at [this](https://github.com/HofmeisterAn/dotnet-testcontainers/blob/develop/src/DotNet.Testcontainers/Core/Models/Database/PostgreSqlTestcontainerConfiguration.cs) and [this](https://github.com/HofmeisterAn/dotnet-testcontainers/blob/0f05e0d9975d5c3482ed767f498ae1239ebd9fe2/src/DotNet.Testcontainers/Core/Builder/TestcontainersBuilderDatabaseExtension.cs), maybe the links will help you to implement the missing parts. Maybe you are interested to contribute, any help is welcome. If you like, I can guide you.

**Comment by emzyme20** [comment]
@HofmeisterAn Hi there, I am this week looking at modifying the open source project to support different host ports to use in the PostgreSqlContainer.

I'm not sure changing the ConnectionString will work as that is what I attempted to do by overriding it in my test instance where the container was created..

What methods do you think are missing?

**Comment by emzyme20** [comment]
If I change the code in your PostgreSqlContainer here:

```
        public override string ConnectionString
        {
            get
            {
               (int _, int portBinding) = PortBindings.First();
                
                return $"Host={GetDockerHostIpAddress()};Port={portBinding};Username={UserName};pwd={Password}";                    
            }
        }
```

Your simple test will pass if I supply it with a random port. The issue is that the PortBindings that you can supply is an array of bindings.. By taking .First() seems a bit of a smell to me as it then does not make sense to be an Array.

**Comment by HofmeisterAn** [comment]
> What methods do you think are missing?

I don't think other database and user names are supported right now. You used `root` and `TestDb` in you example. The required env variables `POSTGRES_USER` and `POSTGRES_DB` are not set at the moment. See for comparison `POSTGRES_PASSWORD`. Also have a look at the section [Environment Variables](https://hub.docker.com/_/postgres).

> The issue is that the PortBindings that you can supply is an array of bindings...

Yes, `Port` should be a property for `DatabaseContainer`. The configured ports should be mapped to the default database port.

**Comment by emzyme20** [comment]
Sorry for the confusion. I only set the database name and username as we were debugging this issue. In the very first post, I was just using the defaults and it was still clashing with the ports.

I don't think I understand how I can map and array of PortBindings to a single integer (unless you are happy with my use of .First())

**Comment by emzyme20** [comment]
I've forked and made a change. Have a look and let me have your thoughts.

[Pull Request for Issue](https://github.com/testcontainers/testcontainers-dotnet/pull/55)

**Comment by HofmeisterAn** [comment]
@emzyme20 I'm busy today, and I'll get back later to you. But `Port` should be a new property in `DatabaseContainer` like `Username` or `Password`. Then `PortBindings.First();` is unnecessary.

**Comment by emzyme20** [comment]
@HofmeisterAn - I've updated my Pull request branch to use the method that you have asked for. Hope it's ok now?

**Comment by emzyme20** [comment]
I built locally and referenced the new binaries in my project and the unit tests now all run in parallel. Do you think my changes can get merged and released soon?

**Comment by emzyme20** [comment]
@HofmeisterAn Hi, I was just wondering if you would be able to let me know a planned release/fix for this?

**Comment by swissarmykirpan** [comment]
Hi @emzyme20 as discussed in the PR, we've decided to go down the route of enabling `PublishAllPorts`. This change has been submitted to nuget.org and should be available very shortly!

Thank you for your patience and I hope you will continue to contribute to pushing this forwards. :)

**Comment by emzyme20** [comment]
Where should issues be created now for this issue? It doesn't seem to be working, I wasn't sure if I should use the old pull request or this issue or potentially open a new issue?:

@swissarmykirpan, @HofmeisterAn 

I've checked out your latest beta release: 0.0.2.107

However, when I drop the line that you mentioned (.WithPortBindings), I now get errors because the container has not started and you are trying to request the port. I can see from the code difference link that you have now added that exception to the GetMappedPort().

What's the new way to initialise the container and get the connection string?

I can't seem to access the `Container.ConnectionString` property at all without it throwing an exception with:

> 'this.Container.ConnectionString' threw an exception of type 'System.InvalidOperationException'
> 
> Container must be started before mapped ports can be retrieved

If I construct my own connection string property, it also fails. It seems that any call to your new method GetMappedPort will hit the first exception block inside. I need to have access to the connection string to be able to prepare the initialisation code for my DocumentStore.

My full code that no longer works with this version:

``` csharp
    public class PostgreSqlFixture : IAsyncLifetime
    {
        private readonly int hostPort;

        public PostgreSqlFixture()
        {
            this.hostPort = FreeTcpPort();
            this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();

            this.DocumentStore = global::Marten.DocumentStore.For(
                config =>
                {
                    config.Connection(this.ConnectionString);
                });
        }

        public string ConnectionString => $"{this.Container.GetDockerHostIpAddress()}:{this.Container.GetMappedPort(this.hostPort)}";

        public IDocumentStore DocumentStore { get; }

        private PostgreSqlContainer Container { get; }

        public Task DisposeAsync()
        {
            return this.Container.Stop();
        }

        public Task InitializeAsync()
        {
            return this.Container.Start();
        }

        private static int FreeTcpPort()
        {
            var l = new TcpListener(IPAddress.Loopback, 0);
            l.Start();
            var port = ((IPEndPoint)l.LocalEndpoint).Port;
            l.Stop();

            return port;
        }
    }
```

**Comment by emzyme20** [comment]
Ok, I can get it to work, but I had to pull all the document store initialisation out of the constructor into a property. That way the starting of the container will have happened. It will do and it's a workaround that seems ok

``` csharp
    public class PostgreSqlFixture : IAsyncLifetime
    {
        public PostgreSqlFixture()
        {
            this.Container = new DatabaseContainerBuilder<PostgreSqlContainer>()
                .Begin()
                .WithImage($"{PostgreSqlContainer.IMAGE}:{PostgreSqlContainer.DEFAULT_TAG}")
                .WithExposedPorts(PostgreSqlContainer.POSTGRESQL_PORT)
                .WithEnv(("POSTGRES_PASSWORD", "Password123"))
                .Build();
        }

        public IDocumentStore DocumentStore
        {
            get
            {
                return global::Marten.DocumentStore.For(
                    config =>
                    {
                        config.Connection(this.Container.ConnectionString);
                    });
            }
        }

        private PostgreSqlContainer Container { get; }

        public Task DisposeAsync()
        {
            return this.Container.Stop();
        }

        public Task InitializeAsync()
        {
            return this.Container.Start();
        }
    }
```
